django/urls/base.py

```diff
@@ -85,7 +94,16 @@ def reverse(viewname, urlconf=None, args=None, kwargs=None, current_app=None):
                 ns_pattern, resolver, tuple(ns_converters.items())
             )
 
-    return resolver._reverse_with_prefix(view, prefix, *args, **kwargs)
+    resolved_url = resolver._reverse_with_prefix(view, prefix, *args, **kwargs)
+    if query:
+        if isinstance(query, QueryDict):
+            query_string = query.urlencode()
+        else:
+            query_string = urlencode(query, doseq=True)
+        resolved_url += "?" + query_string
+    if fragment:
+        resolved_url += "#" + fragment
```

Comment:
Was going to ask whether we need to encode the fragment, though I'm not sure it really matters as the browser seems to alter it as need be.

Interestingly:
 - Chrome will %-encode ie if you manually type "Section 2" it will replace it with "Section%202"
 - Firefox is the exact opposite; it replaces "Section%202" with "Section 2"

Reply:
I think leaving it alone is fine—that's the behaviour of `ParseResult` too, which I would expect to do the "right thing".

I think it would be a challenge to encode it correctly anyway, as the encoding of the fragment should [depend on the MIME type of the resource being linked to](https://stackoverflow.com/a/20681028), which Django can't really know.